ansible_events_ui/db/utils/lostream.py

```diff
@@ -0,0 +1,227 @@
+from __future__ import annotations
+
+import logging
+from io import UnsupportedOperation
+from typing import Tuple, Union
+
+from fastapi import Depends
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from ansible_events_ui.db.dependency import get_db_session
+
+LOG = logging.getLogger(__name__)
+
+
+# Default read buffer size is
+CHUNK_SIZE = 1024**2
+
+
+class LObject:
+    """
+    Facilitate PostgreSQL large object interface using server-side functions.
+
+    As of 2022-09-01, there is no large object support directly
+    in asyncpg or other async Python PostgreSQL drivers.
+    """
+
+    VALID_MODES = {"rwt", "rt", "wt", "at", "rwb", "rb", "wb", "ab"}
+    INV_READ = 0x00040000
+    INV_WRITE = 0x00020000
+    MODE_MAP = {"r": INV_READ, "w": INV_WRITE, "a": INV_READ | INV_WRITE}
+
+    def __init__(
+        self,
+        oid: int = 0,
+        length: int = 0,
+        session: AsyncSession = None,
+        *,
+        chunk_size: int = CHUNK_SIZE,
+        mode: str = "rb",
+    ):
+        self.session = session
+        self.chunk_size = chunk_size if chunk_size > 0 else CHUNK_SIZE
+        self.oid = oid
+        self.length = length
+        self.closed = False
+
+        self.imode, self.text_data, self.append = self._resolve_mode(mode)
+        self.pos = self.length if self.append else 0
+
+    def _resolve_mode(self, mode: str):
+        if mode not in self.VALID_MODES:
+            raise ValueError(
+                f"Mode {mode} must be one of {sorted(self.VALID_MODES)}"
+            )
+
+        _text_data = mode.endswith("t")
+        _append = mode.startswith("a")
+        _mode = 0
+        for c in mode:
+            _mode |= self.MODE_MAP.get(c, 0)
+
+        return _mode, _text_data, _append
+
+    def closed_check(self):
+        if self.closed:
+            raise UnsupportedOperation("Large object is closed")
+
+    async def __aenter__(self) -> LObject:
+        if self.closed:
+            raise UnsupportedOperation("Large object is closed")
+        return self
+
+    async def __aexit__(self, *aexit_stuff: Tuple) -> None:
+        await self.close()
+        return None
+
+    async def gread(self) -> Union[bytes, str]:
+        self.pos = 0
+        buff = b"\x00"
+        while len(buff) > 0:
+            buff = await self.read()
+            yield buff
+
+    async def read(self) -> Union[bytes, str]:
+        self.closed_check()
+        if self.imode == self.INV_WRITE:
+            raise UnsupportedOperation("not readable")
+
+        sql = """
+select lo_get(:_oid, :_pos, :_len) as lo_bytes
```

Comment:
Can you please not break indentation level with SQL queries. It makes code harder to read. SQL is not sensitive for extra whitespaces and some queries are short enough to be written as single line string literal.

Reply:
And yet it breaks the indentation flow of a code, which makes code harder to read. Please avoid it.

Reply:
I can see the merits for both, but far more important than which one we go with is that we're consistent. If we're currently doing it the way that @cutwater is suggesting, then we should change these, but if not and we're establishing a new precedent, I would suggest we leave these the way they are for this PR and decide on a standard with input from the rest of the team.

Reply:
I think that code legibility is more important than the cut&paste.

Reply:
@eclarizio here are some examples of existing code:

1. https://github.com/benthomasson/ansible-events-ui/blob/3e47a3387c0404a965c6465aeafb5de286d12f06/ansible_events_ui/db/migrations/versions/202208221331_4c2b57da4a58_project_fkeys.py
2. https://github.com/benthomasson/ansible-events-ui/blob/3e47a3387c0404a965c6465aeafb5de286d12f06/ansible_events_ui/db/migrations/versions/202208221331_4c2b57da4a58_project_fkeys.py
3. https://github.com/benthomasson/ansible-events-ui/blob/3e47a3387c0404a965c6465aeafb5de286d12f06/ansible_events_ui/db/utils/migrations.py

Reply:
@eclarizio 
I accept that I'm breaking with the convention that has been established in other parts of the project, but, in my defense:

1. We do not use raw SQL that often and we are apparently planning to rewrite all of the models and queries (unless I'm misunderstanding the 2.0 query style ticket(s))... so the existing raw SQL could be rewritten to utilize `<model>.query` or `<model>(params).save()` or `<model>.query.filter().update()`. forms and the raw SQL in the examples eliminated altogether.
2. The SQL in the `lostream` module is **required** (at least, for now) because none of the async postgresql drivers support a direct large object interface.
3. Right-justifying the SQL code, in my opinion, becomes difficult to maintain with editors that enforce tab-stops and also constrict how much code can be placed on a single line which can impact how readable a query is in the first place.
4. I don't foresee at this time much more raw SQL being included in the project since SQLAlchemy is being utilized in other modules. Creating models for `pg_largeobject_metatata` and `pg_largeobject` are unnecessary as they will _only_ be utilized by the `lostream` module.

Reply:
@Red-HAP 

1. Allow me to give more details on the topic. We already use SQLAlchemy 2.0 style in the project, but we don't enforce it at the moment (`future=True` in the engine configuration), so the database engine still runs in the compatibility mode which has left since I've migrated the codebase from the legacy `databases` wrapper.
   
    The `<model>.query` (which is actually `session.query`) is deprecated since SQLAlchemy 1.4, which introduced new 2.0-style. You can find more details on migration here: https://docs.sqlalchemy.org/en/14/changelog/migration_20.html
   
    The declarative style, is a different story completely. The intention is to use ORM capabilities to utilize data mapper facility and model relationship configuration, which will simplify querying and working with related tables.

2. In most cases the raw SQL is used to call functions (i.e. `lo_get`, `lo_put`, `lo_truncate`, etc.). Isn't it possible to use SQLAlchemy for generating those queries, like I proposed in my other comment:

   ```
   sa.select(sa.func.lo_get(self.oid, self.pos, self.chunk_size))
   ```

   There is one raw SQL query remains in `_verify_large_object`, and referring to item nr. 4, I don't think the change is actually needed ( no need to introduce a model for it ).

3. Can you please elaborate on this or give an example of such editors?

    Considering the previous point, if we're talking about single raw SQL query, I can suggest moving it into a module level constant, which has indentation 0:

    ```
    VERIFY_LARGE_OBJECT_QUERY = """
    select m.oid,
           sum(length(coalesce(d.data, convert_to('', 'utf-8')))) as _length
      from pg_largeobject_metadata m
      left
      join pg_largeobject d
        on d.loid = m.oid
     where m.oid = :_oid
     group
       by m.oid;
    """
    ```

4. See nr. 2